### PR TITLE
[tune] add budgeted hyperparameter tuning algorithm

### PR DESCRIPTION
**Pull request by zhiyun**
### add BudgetedScheduler to ray/tune
BudgetedScheduler suspends and resumes trials based on a budgeted hyperparameter
optimization algorithm. It uses a Bayesian belief model to predict the
future performance of a configuration, and selects the next trial based
on the future prediction and the remaining budget.
### tune_resnet_budget.py 
example to use the BudgetedScheduler.

cc @mitar

### CODE REVIEW

File: python/ray/tune/budget/bayesian_util.py [hunk 1]
```diff
@@ -0,0 +1,437 @@
+from __future__ import division, print_function
+import copy
+import random
+import numpy as np
+import kernel_util as kernel
```
Comment (richardliaw) [inline python/ray/tune/budget/bayesian_util.py]: Generally, we import from the project (ie, `ray.tune.budget.kernel_util`).

File: python/ray/tune/budget/budget.py [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+import collections
+import numpy as np
+
+from ray.tune.trial_scheduler import FIFOScheduler, TrialScheduler
+from ray.tune.trial import Trial
+
+from bayesian_util import BeliefModel
+from acqn_util import longreg
+
+np.warnings.filterwarnings('ignore')
```
Comment (richardliaw) [inline python/ray/tune/budget/budget.py]: We can remove this, right?

File: python/ray/tune/budget/tune_resnet_budget.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
+#!/usr/bin/env python
```
Comment (richardliaw) [inline python/ray/tune/budget/tune_resnet_budget.py]: Let's put this into `ray/tune/examples`

File: python/ray/tune/budget/kernel_util.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+      K[i, j] = m(K[i, j])
+  K *= scale
+  return K
+
+def ft_K_t_t(t, t_star, scale, alpha, beta, tscale=1.0):
+  """Exponential decay mixture kernel"""
+
+  t = np.array(t)
+  t_star = np.array(t_star)
+  assert t.ndim == 1 or (t.ndim == 2 and t.shape[1] == 1)
```
Comment (richardliaw) [inline python/ray/tune/budget/kernel_util.py]: Can you make sure assertions have appropriate warning messages?

File: python/ray/tune/budget/budget.py [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+    def debug_string(self):
+        """This provides a progress notification for the algorithm.
+
+        For each bracket, the algorithm will output a string as follows:
+
+            Bracket(Max Size (n)=5, Milestone (r)=33, completed=14.6%):
+            {PENDING: 2, RUNNING: 3, TERMINATED: 2}
+
+        "Max Size" indicates the max number of pending/running experiments
+        set according to the Hyperband algorithm.
```
Comment (richardliaw) [inline python/ray/tune/budget/budget.py]: This is left over from Hyperband right?

File: python/ray/tune/budget/acqn_util.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+  return (tau - mu) * norm.cdf(z) + std * norm.pdf(z)
+
+
+def longreg(ymu, ys2, method):
+  sidx = np.argsort(ymu)
+  bidx = sidx[0]
+  y0 = ymu[bidx]
+  s0 = ys2[bidx]
+
+  if method is 'th1':
```
Comment (richardliaw) [inline python/ray/tune/budget/acqn_util.py]: `method == "thi"`

File: python/ray/tune/budget/bayesian_util.py [hunk 7]
```diff
@@ -0,0 +1,437 @@
...
+    # TODO - use uniform for now
+    # ll += scipy.stats.norm.logpdf(np.log(t_kernel_params['a']))
+    # ll += scipy.stats.norm.logpdf(np.log(t_kernel_params['b']))
+    # ll += np.log(1 / t_kernel_params['scale'])
+    # ll -= t_kernel_params['log_noise']  # Zero loving prior
+
+    return ll
+
+  def ft_posterior(self, x, t, y, t_star):
+    """Freeze thaw posterior (predictive)
```
Comment (richardliaw) [inline python/ray/tune/budget/bayesian_util.py]: Is there some reference for the algorithm implemented here?

File: python/ray/tune/budget/bayesian_util.py [hunk 8]
```diff
@@ -0,0 +1,437 @@
...
+    if ii >= burn:
+      samples.append(copy.deepcopy(oldhyp))
+      print('newhyp = %s' % " ".join(["%.2f" % (oldhyp[s]) for s in range(len(oldhyp))]))
+  return samples, ever_fail
+
+
+### testing
+if __name__ == "__main__":
+
+  # Set up freeze thaw parameters
```
Comment (richardliaw) [inline python/ray/tune/budget/bayesian_util.py]: We can probably use this in unit tests

**Review by richardliaw — CHANGES_REQUESTED** [review]
This is really neat.

Here's a general list of TODOs/requests (in no particular order):

 - [ ] Can you rename helper functions so that they're more accessible?
 - [ ] Documenting helper functions in the util files
 - [ ] Removing leftover comments from testing
 - [ ] Linting
 - [ ] Adding docstrings for public classes (we follow [Google style](http://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html))

 - [ ] Adding unit tests for the Belief Model
 - [ ] Adding documentation in `ray/docs`

**Comment by richardliaw** [comment]
@ericl What are thoughts on introducing a `contrib` module like in TF?

**Comment by mitar** [comment]
(I have not looked at the code and I am basing this only on old  discussion with @zhiyun.)

@zhiyun, have you find a way to run this in parallel? If I understand correctly, your original algorithm/design assumed you are running only one thing at the time, and after each unit of work, deciding which one to do next? I think Ray Tune is great because it allows easy parallelism. So I think there are two questions to me here:

* How to make your algorithm parallel?
* Or, how to make sure Ray Tune does not try to parallelize it.

If anything of this is redundant or I am completely missing something, please ignore. :-)

**Comment by ericl** [comment]
@richardliaw I think having this under a subdir is fine. Actually we can make an schedulers/ dir to clean things up so this can be under schedulers/budget/.

**Comment by ericl** [comment]
[closing stale issue]

**Pull request closed without merging**